R/bootstrap-model.R

```diff
@@ -0,0 +1,628 @@
+#' @export
+new_bootstrap_run <- function(
+    .mod,
+    .suffix = glue("{get_model_id(.mod)}_boot"),
+    .inherit_tags = TRUE,
+    .overwrite = FALSE,
+    increment = FALSE,
+    remove_cov = TRUE,
+    remove_tables = TRUE
+){
+
+  checkmate::assert_class(.mod, NM_MOD_CLASS)
+
+  model_dir <- get_model_working_directory(.mod)
+
+  # Increments the bootstrap run
+  # (i.e. impossible to overwrite an existing run via this function)
+  if(isTRUE(increment)){
+    new_run_id <- get_next_boot_run(model_dir, .suffix)
+    boot_run_id <- paste0(.suffix, "_run_", new_run_id)
+  }else{
+    new_run_id <- 1
+    boot_run_id <- paste0(.suffix)
+  }
+
+
+  boot_run <- copy_model_from(
+    .parent_mod = .mod,
+    .new_model = boot_run_id,
+    .add_tags = "BOOTSTRAP_SUBMISSION",
+    .inherit_tags = .inherit_tags,
+    .update_model_file = TRUE,
+    .overwrite = .overwrite
+  )
+
+  boot_run[[YAML_MOD_TYPE]] <- "nmboot"
+  boot_run <- save_model_yaml(boot_run)
+
+  # Change problem statement
+  prob <- glue("Bootstrap run {new_run_id} of model {get_model_id(.mod)}")
+  modify_prob_statement(boot_run, prob)
+
+  # Optionally remove $TABLE and $COV statements here
+  if(isTRUE(remove_cov)) remove_records(boot_run, type = "covariance")
+  if(isTRUE(remove_tables)) remove_records(boot_run, type = "table")
+
+  # Return read-in model to get the updated class as part of the model object
+  return(read_model(file.path(model_dir, boot_run_id)))
+}
+
+
+get_next_boot_run <- function(model_dir, .suffix){
+  model_dir_files <- fs::dir_ls(model_dir, glob = "*.yaml")
+  boot_runs <- model_dir_files[grepl(.suffix, model_dir_files)]
+  if(!rlang::is_empty(boot_runs)){
+    boot_runs <- fs::path_ext_remove(basename(boot_runs))
+    run_ids <- gsub(.suffix, "", boot_runs)
+    run_ids <- readr::parse_number(run_ids) %>% suppressWarnings()
+    run_ids[is.na(run_ids)] <- 1 # assign 1 if no `_run_[x]` is found
+    new_run <- max(run_ids) + 1
+  }else{
+    new_run <- 1
+  }
+  return(new_run)
+}
+
+
+#' Set up a bootstrap model run.
+#'
+#' Creates a new model object and re-sampled dataset per model run.
+#'
+#' @param .boot_run a `bbi_nmboot_model` object.
+#' @param n number of unique sampled keys, defaults to match dataset
+#' @param strat_cols columns to maintain proportion for stratification
+#' @param replace whether to stratify with replacement
+#' @param seed a seed for sampling the data
+#' @param .overwrite logical (T/F) indicating whether or not to overwrite existing
+#'  setup for a bootstrap run.
+#'
+#' @seealso new_bootstrap_run
+#'
+#' @examples
+#' \dontrun{
+#'
+#' # Setup
+#' .boot_run <- new_bootstrap_run(.mod)
+#' .boot_run <- setup_bootstrap_run(.boot_run, n = 1000, seed = 1234)
+#'
+#' # Submit
+#' submit_model(.boot_run)
+#'
+#' # Check status of runs during submission
+#' get_model_status(.boot_run)
+#' }
+#' @export
+setup_bootstrap_run <- function(
+    .boot_run,
+    n = 100,
+    strat_cols = NULL,
+    replace = TRUE,
```

Comment:
Drop the replace functionality; for bootstrap applications, it should always be true.